dali/python/nvidia/dali/pipeline.py

```diff
@@ -311,30 +357,51 @@ def build(self):
         if self._built:
             return
 
+        if self.num_threads < 1:
+            raise ValueError("Pipeline created with `num_threads` < 1 can only be used "
+                             "for serialization.")
+
         if not self._prepared:
-            self._prepare_graph()
+            self._prepare_graph(define_graph)
 
         self._pipe.Build(self._names_and_devices)
         self._built = True
 
-    def feed_input(self, ref, data, layout=""):
-        """Bind the NumPy array to a tensor produced by ExternalSource
-        operator. It is worth mentioning that `ref` **should not** be overridden
-        with other operator outputs and it should be called from the
-        inside of `iter_setup` method"""
+    def feed_input(self, data_node, data, layout=""):
+        """Bind a NumPy array (or a list thereof) to an output of ExternalSource.
+
+        `data_node` : DataNode or str
+        The `DataNode` returned by a call to ExternalSource or a name of the ExternalSource
+
+        `data` : numpy.ndarray or a list thereof
+        The data to be used as the output of the ExternalSource referred to by `data_node`.
+        In case of GPU external sources, this must be a numpy.ndarray.
+
+        `layout`: str
+        The description of the data layout (or empty string, if not specified).
+        It should be a string of the length that matches the dimensionality of the data, batch
+        dimension excluded. For a batch of channel-first images, this should be "CHW", for
+        channel-last video it's "FHWC" and so on.
+        """
```

Comment:
Nice doc, but shouldn't we use TensorLayout instead of `str`? I know it's technically the string, but if you use TensorLayout you can get linked to the docs about it, and I asked @jantonguirao to update them in his PR about layouts.

Reply:
No :)